### PR TITLE
Add deployment pipeline dev

### PR DESCRIPTION
**Pull request by merrilmathew**
### Context

This PR adds the CICD workflow and the deployment configuration to deploy access-your-teaching-profile development application on Azure app service. 

### Link to Trello card

https://trello.com/c/FDJp2OVb

### CODE REVIEW

**Comment by merrilmathew** [comment]
[Deploy dev environment](https://trello.com/c/FDJp2OVb/215-deploy-dev-environment)

**Review by tvararu — APPROVED** [review]

File: .github/workflows/actions/deploy-environment/action.yml [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  environment_name:
+    description: "The name of the environment"
+    required: true
+  image_tag:
+    description: "github SHA of the image"
+    required: true
+  azure_credentials:
+    description: "JSON object containing a service principal that can read from Azure Key Vault"
+    required: true
+  terraform_vars:
```
Comment (saliceti) [inline .github/workflows/actions/deploy-environment/action.yml]: terraform_vars can be determined with the environment name. can we remove the input?

File: .github/workflows/actions/deploy-environment/action.yml [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    description: "Path to the tfvars file for the environment"
+    required: true
+  terraform_backend_vars:
```
Comment (saliceti) [inline .github/workflows/actions/deploy-environment/action.yml]: Is this input used?

File: .github/workflows/actions/deploy-environment/action.yml [hunk 3]
```diff
@@ -0,0 +1,128 @@
+name: Deploy to Azure
+
+inputs:
+  environment_name:
+    description: "The name of the environment"
```
Comment (saliceti) [inline .github/workflows/actions/deploy-environment/action.yml]: Could we remove all the extra double quotes in the yaml files?

File: .github/workflows/database-backup.yml [hunk 4]
```diff
@@ -0,0 +1,119 @@
+name: Backup Database to Azure Storage
```
Comment (saliceti) [inline .github/workflows/database-backup.yml]: Should this be part of another PR?

File: Dockerfile [hunk 5]
```diff
@@ -56,9 +56,10 @@ RUN rm -rf node_modules log/* tmp/* /tmp && \
 
 # Build runtime image
 FROM ruby:3.1.2-alpine as production
+ARG GIT_SHA
```
Comment (saliceti) [inline Dockerfile]: Please move ARG and ENV GIT_SHA at the very end, to optimise caching

File: .github/workflows/actions/deploy-environment/action.yml [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+      shell: bash
+      env:
+        TFVARS: ${{ inputs.terraform_vars }}
+      working-directory: terraform
+
+    - uses: Azure/login@v1
+      with:
+        creds: ${{ inputs.azure_credentials }}
+
+    - uses: Azure/get-keyvault-secrets@v1
```
Comment (smneal) [inline .github/workflows/actions/deploy-environment/action.yml]: Is it worth changing this to use the [azure/cli@v1](https://github.com/Azure/cli) action as the [get-keyvault-secrets](https://github.com/Azure/get-keyvault-secrets) action is deprecated now? @RMcVelia I think you recently had to do this?

File: .github/workflows/actions/deploy-environment/action.yml [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+      id: get_secrets
+      with:
+        keyvault: ${{ steps.config.outputs.key_vault_name }}
+        secrets: "TFSTATE-CONTAINER-ACCESS-KEY"
+
+    - uses: hashicorp/setup-terraform@v1
```
Comment (smneal) [inline .github/workflows/actions/deploy-environment/action.yml]: This task has a `v2` now.

File: .github/workflows/database-backup.yml [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+          echo "POSTGRES_SERVER_HOST_NAME=${{ env.RESOURCE_PREFIX }}aytp-${{ env.ENV }}-psql.postgres.database.azure.com" >> $GITHUB_ENV
+          echo "POSTGRES_DATABASE_NAME=access_your_teaching_profile_production" >> $GITHUB_ENV
+
+      # - uses: Azure/get-keyvault-secrets@v1
+      #   id: get_storage_secret
+      #   with:
+      #     keyvault: ${{ env.KEY_VAULT_NAME }}
+      #     secrets: "BACKUP-STORAGE-CONNECTION-STRING"
+
+      - uses: DfE-Digital/keyvault-yaml-secret@v1
```
Comment (smneal) [inline .github/workflows/database-backup.yml]: @saliceti Did this get a version bump to fix some issues?

File: Makefile [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+	[[ "${SP_AUTH}" != "true" ]] && az account set -s $(AZURE_SUBSCRIPTION) || true
+	terraform -chdir=terraform init -backend-config workspace_variables/${DEPLOY_ENV}.backend.tfvars $(backend_config) -upgrade -reconfigure
+
+terraform-plan: terraform-init
+	terraform -chdir=terraform plan -var-file workspace_variables/${DEPLOY_ENV}.tfvars.json
+
+terraform-apply: terraform-init
+	terraform -chdir=terraform apply -var-file workspace_variables/${DEPLOY_ENV}.tfvars.json ${AUTO_APPROVE}
+
+terraform-apply-replace-redis: terraform-init # make dev terraform-apply-replace-redis PASSCODE="XXX"
```
Comment (saliceti) [inline Makefile]: Is this really required?

File: terraform/app.tf [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+resource "azurerm_postgresql_flexible_server_firewall_rule" "postgres-fw-rule-azure" {
+  name             = "AllowAzure"
+  server_id        = azurerm_postgresql_flexible_server.postgres-server.id
+  start_ip_address = "0.0.0.0"
+  end_ip_address   = "0.0.0.0"
+}
+
+resource "azurerm_redis_cache" "redis" {
```
Comment (saliceti) [inline terraform/app.tf:53]: Can you confirm if redis is required?

File: terraform/app.tf [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+  identity {
+    type = "SystemAssigned"
+  }
+
+  site_config {
+    http2_enabled       = true
+    minimum_tls_version = "1.2"
+    health_check_path   = "/health"
+    # application_stack {
```
Comment (saliceti) [inline terraform/app.tf]: Can this be removed?

File: terraform/provider.tf [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+  subscription_id            = try(local.azure_credentials.subscriptionId, null)
+  client_id                  = try(local.azure_credentials.clientId, null)
+  client_secret              = try(local.azure_credentials.clientSecret, null)
+  tenant_id                  = try(local.azure_credentials.tenantId, null)
+  skip_provider_registration = true
+
+  features {}
+}
+
+provider "statuscake" {
```
Comment (saliceti) [inline terraform/provider.tf]: statuscake is not used right now. can we add this in another card/PR?

File: terraform/terraform.tf [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+    container_name = "aytp-tfstate"
+  }
+
+  required_providers {
+    azurerm = {
+      source  = "hashicorp/azurerm"
+      version = "~> 3.23.0"
+    }
+
+    azapi = {
```
Comment (saliceti) [inline terraform/terraform.tf]: The azapi provider is not used

File: terraform/terraform.tf [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+  required_version = "~> 1.0"
+
+  backend "azurerm" {
+    container_name = "aytp-tfstate"
+  }
+
+  required_providers {
+    azurerm = {
+      source  = "hashicorp/azurerm"
+      version = "~> 3.23.0"
```
Comment (saliceti) [inline terraform/terraform.tf]: Can we use the latest version? 3.25.0

File: terraform/variables.tf [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+  type    = string
+  default = "PerGB2018"
+}
+
+variable "docker_image" {
+  type    = string
+  default = "ghcr.io/dfe-digital/access-your-teaching-profile"
+}
+
+# variable "docker_image_tag" {
```
Comment (saliceti) [inline terraform/variables.tf]: Remove comment?

File: terraform/app.tf (hunk 10 above)
Comment (merrilmathew) [reply terraform/app.tf:53]: Yes with + Sidekiq

File: .github/workflows/actions/deploy-environment/action.yml [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+    - name: Get TFSTATE-CONTAINER-ACCESS-KEY
+      run: |
+        TFSTATE_CONTAINER_ACCESS_KEY="$(az storage account keys list -g ${{ steps.config.outputs.resource_group_name }} -n ${{ steps.config.outputs.storage_account_name }} | jq -r '.[0].value')"
+        echo "::add-mask::$TFSTATE_CONTAINER_ACCESS_KEY"
+        echo "TFSTATE_CONTAINER_ACCESS_KEY=$TFSTATE_CONTAINER_ACCESS_KEY" >> $GITHUB_ENV
+      shell: bash
+
+    - uses: hashicorp/setup-terraform@v2
+      with:
+        terraform_version: 1.0.10
```
Comment (saliceti) [inline .github/workflows/actions/deploy-environment/action.yml]: Can we use the latest terraform?

File: terraform/terraform.tf [hunk 17]
```diff
@@ -0,0 +1,14 @@
+terraform {
+  required_version = "~> 1.0"
```
Comment (saliceti) [inline terraform/terraform.tf]: Can we pin it to the latest available version? Without the squiggly arrow to avoid surprises
